base/dcp_receiver.go

```diff
@@ -202,9 +204,90 @@ type NoPasswordAuthHandler struct {
 
 func (nph NoPasswordAuthHandler) GetCredentials() (username string, password string, bucketname string) {
 	_, _, bucketname = nph.Handler.GetCredentials()
+
 	return "", "", bucketname
 }
 
+// getHighSeqMetadata returns metadata to feed into a DCP client based on the last sequence numbers stored in memory
+func getHighSeqMetadata(bucket Bucket) ([]DCPMetadata, error) {
+	store, ok := AsCouchbaseStore(bucket)
+	if !ok {
+		return nil, errors.New("DCP Client requires bucket to be CouchbaseStore")
+	}
+
+	numVbuckets, err := store.GetMaxVbno()
+	if err != nil {
+		return nil, fmt.Errorf("Unable to determine maxVbNo when creating DCP client: %w", err)
+	}
+
+	_, highSeqNos, statsErr := store.GetStatsVbSeqno(numVbuckets, true)
+	if statsErr != nil {
+		return nil, fmt.Errorf("Unable to obtain high seqnos for one-shot DCP feed: %w", statsErr)
+	}
+
+	metadata := make([]DCPMetadata, numVbuckets)
+	for vbNo := uint16(0); vbNo < numVbuckets; vbNo++ {
+		// TODO: should we/can we initialize VbUUID or other fields?
+		metadata[vbNo].FailoverEntries = make([]gocbcore.FailoverEntry, 0)
+		metadata[vbNo].StartSeqNo = gocbcore.SeqNo(highSeqNos[vbNo])
+		metadata[vbNo].EndSeqNo = gocbcore.SeqNo(uint64(0xFFFFFFFFFFFFFFFF))
+	}
+	return metadata, nil
+}
+
+// StartGOCB2DCPFeed starts a DCP Feed.
+func StartGOCB2DCPFeed(bucket Bucket, spec BucketSpec, args sgbucket.FeedArguments, callback sgbucket.FeedEventCallbackFunc, dbStats *expvar.Map) error {
+	metadata, err := getHighSeqMetadata(bucket)
+	if err != nil {
+		return err
+	}
+	bucketName := spec.BucketName
+	// TODO: does this need to be more unique, if two sync gateways connect to same server, will this reject the second DCP feed?
+	feedName := fmt.Sprintf("%s-%s:%s", args.ID, ProductAPIVersion, bucketName)
+	dcpClient, err := NewDCPClient(feedName,
+		callback,
+		DCPClientOptions{
+			MetadataType:    DCPMetadataInMemory,
+			InitialMetadata: metadata},
+		bucket,
+		"")
+	if err != nil {
+		return err
+	}
+
+	doneChan, err := dcpClient.Start()
+	loggingCtx := context.TODO()
+	if err != nil {
+		ErrorfCtx(loggingCtx, "Failed to start DCP Feed [%s]", MD(feedName))
+		_ = dcpClient.Close()
+		args.Started <- err
+		close(args.Started)
+		return err
+	}
+	if args.Started != nil {
+		args.Started <- nil
+		close(args.Started)
+	}
+	InfofCtx(loggingCtx, KeyDCP, "Started DCP Feed [%s]", MD(feedName))
+	go func() {
+		select {
+		case <-doneChan:
+			TracefCtx(loggingCtx, KeyDCP, "Closed DCP Feed [%s-%s].", MD(feedName))
+			err = <-doneChan
+			break
+		case <-args.Terminator:
+			InfofCtx(loggingCtx, KeyDCP, "Closing DCP Feed [%s] based on termination notification", MD(feedName))
+			dcpClient.Close()
+			err = <-doneChan
+			break
+		}
+		if args.DoneChan != nil {
+			close(args.DoneChan)
+		}
+	}()
+	return err
```

Comment:
`err` is only ever assigned to in the above goroutine, which will have likely not started/run to completion before this function returns `err`

I'm not sure how we're handling the errors that come from closing the DCP client in this implementation.